test/unit/test_equation_manager.py

```diff
@@ -0,0 +1,448 @@
+""" Tests of the EquationManager used in the Ad framework.
+
+The tests focus on various assembly methods:
+    * test_secondary_variable_assembly: Test of assembly when secondary variables are
+        present.
+    * test_assemble_subsystem: Assemble blocks of the full set of equations.
+    * test_extract_subsystem: Extract a new EquationManager for a subset of equations.
+    * test_schur_complement: Assemble a subsystem, using a Schur complement reduction.
+
+"""
+import pytest
+import numpy as np
+import porepy as pp
+import scipy.sparse as sps
+
+
+class EquationManagerSetup:
+    # Class to set up an EquationManager with a combination of variables
+    # and equations, designed to make it convenient to test critical functionality
+    # of the EM.
+
+    def __init__(self):
+        g1 = pp.CartGrid([3, 1])
+        g2 = pp.CartGrid([4, 1])
+
+        gb = pp.GridBucket()
+        gb.add_nodes([g1, g2])
+
+        d1 = gb.node_props(g1)
+        d2 = gb.node_props(g2)
+
+        for g, d in gb:
+            d[pp.PRIMARY_VARIABLES] = {
+                "x": {"cells": 1},
+                "y": {"cells": 1},
+            }
+            if g == g2:
+                # Add a third variable on the second
+                d[pp.PRIMARY_VARIABLES].update({"z": {"cells": 1}})
+
+            x = 2 * np.ones(g.num_cells)
+            y = 3 * np.ones(g.num_cells)
+            z = 4 * np.ones(g.num_cells)
+
+            d[pp.STATE] = {
+                "x": x,
+                "y": y,
+                "z": z,
+                pp.ITERATE: {"x": x.copy(), "y": y.copy(), "z": z.copy()},
+            }
+
+        # Ad representation of variables
+        dof_manager = pp.DofManager(gb)
+        eq_manager = pp.ad.EquationManager(gb, dof_manager)
+
+        x_ad = eq_manager.variable(g2, "x")
+        y_ad = eq_manager.variable(g2, "y")
+        z_ad = eq_manager.variable(g2, "z")
+
+        x_merged = eq_manager.merge_variables([(g1, "x"), (g2, "x")])
+        y_merged = eq_manager.merge_variables([(g1, "y"), (g2, "y")])
+
+        projections = pp.ad.SubdomainProjections(grids=[g1, g2])
+        proj = projections.cell_restriction(g2)
+
+        # One equation with only simple variables (not merged)
+        eq_simple = x_ad * y_ad + 2 * z_ad
+        # One equation using only merged variables
+        eq_merged = y_merged * (1 + x_merged)
+        # One equation combining merged and simple variables
+        eq_combined = x_ad + y_ad * (proj * y_merged)
+
+        eq_manager.equations.update(
+            {"simple": eq_simple, "merged": eq_merged, "combined": eq_combined}
+        )
+
+        A, b = eq_manager.assemble()
+        self.A = A
+        self.b = b
+        self.eq_manager = eq_manager
+
+        self.x1 = x_ad
+        self.x2 = eq_manager.variable(g1, "x")
+
+        self.z1 = z_ad
+        self.x_merged = x_merged
+        self.y_merged = y_merged
+
+        self.g1 = g1
+        self.g2 = g2
+
+    ## Helper methods below.
```

Comment:
Just to be sure: Is this our preferred way of distinguishing between different "code segments"?